subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -993,6 +992,22 @@ public DependencySet getAllDependencies() {
         return allDependencies;
     }
 
+    @Override
+    public boolean hasDependencies() {
+        return !getAllDependencies().isEmpty();
+    }
+
+    /**
+     * This method is a heuristic that gives an idea of the "size" of the graph. The larger
+     * the graph is, the higher the risk of internal resizes exists, so we try to estimate
+     * the size of the graph to avoid maps resizing.
+     */
+    @Override
+    public int getEstimatedGraphSize() {
+        int estimate = (int) (512 * Math.log(getAllDependencies().size()));
+        return Math.max(10, estimate);
```

Comment:
Should have a comment explaining why these specific numbers and operations are used.

Reply:
Hmm, I see this was just copied from elsewhere. Still, it would be nice to explain it.

Reply:
Looking at git history, there is no reasoning why these specific numbers or operations were chosen. I can add a comment stating that but I can't otherwise explain why it is the way it is.

Reply:
I filed https://github.com/gradle/gradle/issues/24641 to do this at a later date. It may be worth linking to that as a `TODO`.